banditpylib/protocols/collaborative_protocol.py

```diff
@@ -0,0 +1,138 @@
+
+  @property
+  def name(self) -> str:
+    return 'collaborative_learning_protocol'
+
+  def _one_trial(self, random_seed: int, debug: bool) -> bytes:
+    if debug:
+      logging.set_verbosity(logging.DEBUG)
+    np.random.seed(random_seed)
+
+    # initialization
+    current_learner = cast(CollaborativeBAILearner, self.current_learner)
+    current_learner.reset()
+    agents = current_learner.get_agents()
+    bandits = []
+    master = current_learner.get_master()
+    for _ in range(len(agents)):
+      bandits.append(dcopy(self.bandit))
+      bandits[-1].reset()
+
+    trial = Trial()
+    trial.bandit = self.bandit.name
+    trial.learner = current_learner.name
+
+    communication_rounds, total_pulls = 0, 0
+    active_agent_ids = list(range(len(agents)))
+    agent_arm_assignment = master.initial_arm_assignment()
+    for agent_id in agent_arm_assignment:
+      agents[agent_id].set_input_arms(agent_arm_assignment[agent_id])
+
+    while True:
+      max_pulls = 0
+      agent_in_wait_ids = []
+
+      # preparation and learning
+      for agent_id in active_agent_ids:
+        agent = agents[agent_id]
+        pulls = 0
+        while True:
+          actions = agent.actions(bandits[agent_id].context)
+          for arm_pull in actions.arm_pulls:
+            pulls += arm_pull.times
+
+          if actions.state == Actions.DEFAULT_NORMAL:
+            feedback = bandits[agent_id].feed(actions)
+            agent.update(feedback)
+          elif actions.state == Actions.WAIT:
+            agent_in_wait_ids.append(agent_id)
+            break
+          else: # actions.state == Actions.STOP
+            break
+        max_pulls = max(max_pulls, pulls)
+      total_pulls += max_pulls
+
+      # stop if all agents are in STOP <=> no agents in WAIT
+      if not agent_in_wait_ids:
+        break
+
+      # communication and aggregation
+      # key is arm id and target is a tuple storing average rewards and pulls
+      messages: Dict[int, Tuple[float, int]] = {}
+      for agent_id in agent_in_wait_ids:
+        agent = agents[agent_id]
+
+        # key is arm_id, target is Tuple[em_mean_reward, pulls]
+        # empty dict => arm_id was None
+        message_from_agent = agent.broadcast()
+        for arm_id in message_from_agent:
+          if arm_id not in messages:
+            messages[arm_id] = (0.0, 0)
+          arm_info = message_from_agent[arm_id]
+          new_pulls = messages[arm_id][1] + arm_info[1]
+          new_em_mean_reward = (messages[arm_id][1] * messages[arm_id][0] + \
+              arm_info[1] * arm_info[0]) / new_pulls
+          messages[arm_id] = (new_em_mean_reward, new_pulls)
+
+      # Send info to master for elimination to get arm assignment for next round
+      # agent_arm_assignment: key is agent_id, value is a list storing arm ids
+      # assigned to this agent
+      agent_arm_assignment = master.elimination(agent_in_wait_ids, messages)
```

Comment:
Can you put the logic of line 107 to line 121 inside the `master.elimination`? `master.elimination` will accept a dictionary whose key is agent id and value is a list of tuples which denote the information of the arms the agent pulls.